File: src/bridge/bridge-responses.ts
/** Standard bridge response envelopes (keys match iOS/Android native modules). */

import { stubConfigurePlayerResponse } from '../modules/normalize-player-config';

export type CatalogSearchPayload = {
  songs: unknown[];
  albums: unknown[];
  artists: unknown[];
  playlists: unknown[];
  stations: unknown[];
  musicVideos: unknown[];
};

export type CatalogChartsPayload = {
  songs: unknown[];
  albums: unknown[];
  playlists: unknown[];
  musicVideos: unknown[];
};

export type LibrarySearchPayload = {
  songs: unknown[];
  albums: unknown[];
  artists: unknown[];
  playlists: unknown[];
  musicVideos: unknown[];
};

export const BridgeResponses = {
  storefront(id: string): { id: string } {
    return { id };
  },

  authorization(
    status: string,
    musicUserToken: string | null | undefined,
  ): { status: string; musicUserToken: string | null | undefined } {
    return { status, musicUserToken };
  },

  subscription(details: {
    canPlayCatalogContent: boolean;
    canBecomeSubscriber: boolean;
    hasCloudLibraryEnabled: boolean;
    isMusicCatalogSubscriptionEligible: boolean;
  }): {
    canPlayCatalogContent: boolean;
    canBecomeSubscriber: boolean;
    hasCloudLibraryEnabled: boolean;
    isMusicCatalogSubscriptionEligible: boolean;
  } {
    return {
      canPlayCatalogContent: details.canPlayCatalogContent,
      canBecomeSubscriber: details.canBecomeSubscriber,
      hasCloudLibraryEnabled: details.hasCloudLibraryEnabled,
      isMusicCatalogSubscriptionEligible: details.isMusicCatalogSubscriptionEligible,
    };
  },

  playbackState(state: {
    playbackRate: number;
    playbackStatus: string;
    playbackTime: number;
    currentSong?: unknown;
  }): Record<string, unknown> {
    const result: Record<string, unknown> = {
      playbackRate: state.playbackRate,
      playbackStatus: state.playbackStatus,
      playbackTime: state.playbackTime,
    };
    if (state.currentSong !== undefined) {
      result.currentSong = state.currentSong;
    }
    return result;
  },

  catalogSearch(result: CatalogSearchPayload): CatalogSearchPayload {
    return {
      songs: result.songs,
      albums: result.albums,
      artists: result.artists,
      playlists: result.playlists,
      stations: result.stations,
      musicVideos: result.musicVideos,
    };
  },

  catalogCharts(result: CatalogChartsPayload): CatalogChartsPayload {
    return {
      songs: result.songs,
      albums: result.albums,
      playlists: result.playlists,
      musicVideos: result.musicVideos,
    };
  },

  librarySearch(result: LibrarySearchPayload): LibrarySearchPayload {
    return {
      songs: result.songs,
      albums: result.albums,
      artists: result.artists,
      playlists: result.playlists,
      musicVideos: result.musicVideos,
    };
  },

  musicVideos(items: unknown[]): { musicVideos: unknown[] } {
    return { musicVideos: items };
  },

  songs(items: unknown[]): { songs: unknown[] } {
    return { songs: items };
  },

  albums(items: unknown[]): { albums: unknown[] } {
    return { albums: items };
  },

  artists(items: unknown[]): { artists: unknown[] } {
    return { artists: items };
  },

  playlists(items: unknown[]): { playlists: unknown[] } {
    return { playlists: items };
  },

  stations(items: unknown[]): { stations: unknown[] } {
    return { stations: items };
  },

  recentlyPlayedResources(items: unknown[]): { recentlyPlayedItems: unknown[] } {
    return { recentlyPlayedItems: items };
  },

  recentItems(items: unknown[]): { items: unknown[] } {
    return { items: items };
  },

  recommendations(items: unknown[]): { recommendations: unknown[] } {
    return { recommendations: items };
  },

  replaySummaries(items: unknown[]): { summaries: unknown[] } {
    return { summaries: items };
  },

  configurePlayer(options: Record<string, unknown>): Record<string, unknown> {
    return stubConfigurePlayerResponse(options);
  },
} as const;

File: android/src/main/java/expo/modules/applemusic/BridgeResponses.kt
package expo.modules.applemusic

/** Standard bridge response envelopes (keys match iOS and TypeScript). */
internal object BridgeResponses {
  fun storefront(id: String): Map<String, Any?> = mapOf("id" to id)

  fun authorization(status: String, musicUserToken: String?): Map<String, Any?> =
    mapOf("status" to status, "musicUserToken" to musicUserToken)

  fun subscription(
    canPlayCatalogContent: Boolean,
    canBecomeSubscriber: Boolean,
    hasCloudLibraryEnabled: Boolean,
    isMusicCatalogSubscriptionEligible: Boolean,
  ): Map<String, Any?> =
    mapOf(
      "canPlayCatalogContent" to canPlayCatalogContent,
      "canBecomeSubscriber" to canBecomeSubscriber,
      "hasCloudLibraryEnabled" to hasCloudLibraryEnabled,
      "isMusicCatalogSubscriptionEligible" to isMusicCatalogSubscriptionEligible,
    )

  fun playbackState(
    playbackRate: Double,
    playbackStatus: String,
    playbackTime: Double,
    currentSong: Map<String, Any?>?,
  ): Map<String, Any?> =
    buildMap {
      put("playbackRate", playbackRate)
      put("playbackStatus", playbackStatus)
      put("playbackTime", playbackTime)
      if (currentSong != null) put("currentSong", currentSong)
    }

  fun catalogSearch(result: AndroidCatalogService.SearchResult): Map<String, Any?> =
    mapOf(
      "songs" to result.songs,
      "albums" to result.albums,
      "artists" to result.artists,
      "playlists" to result.playlists,
      "stations" to result.stations,
      "musicVideos" to result.musicVideos,
    )

  fun catalogCharts(result: AndroidCatalogService.ChartsResult): Map<String, Any?> =
    mapOf(
      "songs" to result.songs,
      "albums" to result.albums,
      "playlists" to result.playlists,
      "musicVideos" to result.musicVideos,
    )

  fun songs(items: List<Map<String, Any?>>): Map<String, Any?> = mapOf("songs" to items)

  fun albums(items: List<Map<String, Any?>>): Map<String, Any?> = mapOf("albums" to items)

  fun artists(items: List<Map<String, Any?>>): Map<String, Any?> = mapOf("artists" to items)

  fun playlists(items: List<Map<String, Any?>>): Map<String, Any?> = mapOf("playlists" to items)

  fun musicVideos(items: List<Map<String, Any?>>): Map<String, Any?> = mapOf("musicVideos" to items)

  fun librarySearch(result: AndroidLibraryService.LibrarySearchResult): Map<String, Any?> =
    mapOf(
      "songs" to result.songs,
      "albums" to result.albums,
      "artists" to result.artists,
      "playlists" to result.playlists,
      "musicVideos" to result.musicVideos,
    )

  fun stations(items: List<Map<String, Any?>>): Map<String, Any?> = mapOf("stations" to items)

  fun recentlyPlayedResources(items: List<Map<String, Any?>>): Map<String, Any?> =
    mapOf("recentlyPlayedItems" to items)

  fun recentItems(items: List<Map<String, Any?>>): Map<String, Any?> = mapOf("items" to items)

  fun recommendations(items: List<Map<String, Any?>>): Map<String, Any?> =
    mapOf("recommendations" to items)

  fun replaySummaries(items: List<Map<String, Any?>>): Map<String, Any?> = mapOf("summaries" to items)
}
